File: word-search.scala
# Solution of https://leetcode.com/problems/word-search/


object WordSearch {
  def exist(board: Array[Array[Char]], word: String): Boolean = {
    for (i <- 0 until board.length){
      for (j <- 0 until board(0).length){
        if(search(board,word,i,j,0)){
          return true
        }
      }
    }
    return false
  }
  
  def search(board : Array[Array[Char]],word : String,i : Int,j: Int,idx: Int) : Boolean = {
    if(idx == word.length) return true
    if ((0 <=i && i<board.length) && (0<=j && j<board(0).length) && board(i)(j) == word(idx)){
      var temp = board(i)(j)
      board(i)(j) = '0'
      var direction = List(List(1,0),List(-1,0),List(0,1),List(0,-1))
      for(m <- 0 until direction.length){
         if(search(board,word,i+direction(m)(0),j+direction(m)(1),idx+1) == true) return true
      }
      board(i)(j) = temp
    }
    return false
  }
}

File: word-search.py
#solution of https://leetcode.com/problems/word-search/

class Solution(object):
    def exist(self, board, word):
        """
        :type board: List[List[str]]
        :type word: str
        :rtype: bool
        """

        for i in range(len(board)):
            for j in range(len(board[0])):
                if self.search(board, word, i, j, 0):
                    return True
        return False

    def search(self, board, word, i, j, idx):

        if idx == len(word): return True

        if 0 <= i < len(board) and 0 <= j < len(board[0]) and board[i][j] == word[idx]:
            temp, board[i][j] = board[i][j], '0'
            direction = ((1, 0), (-1, 0), (0, 1), (0, -1))
            for x, y in direction:
                if self.search(board, word, i + x, j + y, idx + 1): return True
            board[i][j] = temp

        return False
